File: nazarii_bryniarskyi/fibonacci/src/main/java/org/example/service/Fibonacci.java
package org.example.service;


import org.example.model.FibonacciPair;

import java.util.ArrayList;
import java.util.List;


public class Fibonacci {

    private Fibonacci() {}


    public static FibonacciPair border(final int border,
                                       final List<Integer> values) {
        return trueBorder(border, getStartArray(values), 0);
    }

    public static List<Integer> steps(int steps,
                                      final List<Integer> values) {
        return trueSteps(steps, getStartArray(values));
    }

    private static FibonacciPair trueBorder(final int border,
                                            final List<Integer> values,
                                            int counter) {
        final int valuesSize = values.size();
        final int temp = values.get(valuesSize-1) + values.get(valuesSize-2);

        if (temp > border)
            return new FibonacciPair(counter, values);

        values.add(temp);
        counter++;

        return trueBorder(border, values, counter);
    }

    private static List<Integer> trueSteps(int steps,
                                           final List<Integer> values) {
        steps--;
        final int valuesSize = values.size();
        final int temp = values.get(valuesSize-1) + values.get(valuesSize-2);
        values.add(temp);

        return steps == 0
                ? values
                : trueSteps(steps, values);
    }

    private static List<Integer> getStartArray(final List<Integer> values) {
        final int valuesSize = values.size();
        return new ArrayList<>(
                List.of(values.get(valuesSize-1), values.get(valuesSize-2)));
    }

}

File: krenzhel_kate/fibonacci/Fibonacci.cs
public class Fibonacci
{
    public int counter = 0;
    public Fibonacci() { }
    
    public List<int> CalculateFibonacciBorder(double border, List<int> fibonacci)
    {
        int sum = fibonacci[^1] + fibonacci[^2];
        if (sum < border)
        {
            fibonacci.Add(sum);
            counter += 1;
            return CalculateFibonacciBorder(border, fibonacci);
        }
        else
        {
            Console.WriteLine("Count: " + counter);
            return fibonacci;
        }
    }

    public List<int> CalculateFibonacciSteps(double steps, List<int> fibonacci)
    {
        int sum = fibonacci[^1] + fibonacci[^2];
        if (counter < steps)
        {
            fibonacci.Add(sum);
            counter += 1;
            return CalculateFibonacciSteps(steps, fibonacci);
        }
        else
        {
            return fibonacci;
        }
    }   
}
